### DISCUSSION TITLE
Unable to auth by OIDC with Gitlab 2

### DISCUSSION THREAD

**Discussion by bjun626** [outline/outline#3450] [Self Hosting] [answered]
I'm trying to setup login via self-hosted GitLab. But I'm getting an error after auth callback

## Setup
At the moment, I've already setup till GitLab is able to redirect back the callback URL `https://<GITLAB HOST URL>/auth/oidc.callback`

## Version
Self-hosted Docker GitLab `14.9`
Self-hosted Docker Outline `0.63.0`

## GitLab Setup
![image](https://user-images.githubusercontent.com/34598848/164981443-6cbb080e-1d51-4ee7-83e1-adcbf745a214.png)
> I know the most permission at not necessary, but I just want to make sure its not the permission issues

## Auto Redirect Callback
![image](https://user-images.githubusercontent.com/34598848/164981523-f9a142a7-62bb-49cb-8291-e8f1db1954a4.png)

But after the callback I'm still unable to logging due to the following error

![image](https://user-images.githubusercontent.com/34598848/164981481-b552329a-acbd-4fa3-af6e-a15f05540b15.png)

In the outline docker logs after the call back I got the following error message

```
{"label":"http","level":"info","message":" --> POST /auth/oidc 302 2ms 2b"}
{"label":"http","level":"info","message":" <-- GET /auth/oidc.callback?code=cc99<REDACTED>07e38e47512f0e93528b99098fd0dbbfcf5c30b1&state=25<REDACTED>ac7aea52f"}
{"error":"Failed to obtain access token","stack":"InternalOAuthError: Failed to obtain access token\n at OAuth2Strategy._createOAuthError (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:423:17)\n at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:177:45\n at /opt/outline/node_modules/oauth/lib/oauth2.js:191:18\n at ClientRequest.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:162:5)\n at ClientRequest.emit (node:events:526:28)\n at ClientRequest.emit (node:domain:475:12)\n at TLSSocket.socketErrorListener (node:_http_client:442:9)\n at TLSSocket.emit (node:events:526:28)\n at TLSSocket.emit (node:domain:475:12)\n at emitErrorNT (node:internal/streams/destroy:157:8)\n at emitErrorCloseNT (node:internal/streams/destroy:122:3)\n at processTicksAndRejections (node:internal/process/task_queues:83:21)","level":"error","message":"Error during authentication"}

{"label":"http","level":"info","message":" --> POST /auth/oidc.callback?code=cc99<REDACTED>38e47512f0e93528b99098fd0dbbfcf5c30b1&state=259<REDACTED>a52f 302 63ms 69b"}
{"label":"http","level":"info","message":" <-- GET /?notice=auth-error"}

{"label":"http","level":"info","message":" --> GET /?notice=auth-error 200 3ms 1000b"}
```

## Outline ENV Config

```
OIDC_CLIENT_ID=<REDACTED>
OIDC_CLIENT_SECRET=<REDACTED>
OIDC_AUTH_URI=https://<GITLAB HOST URL>/oauth/authorize
OIDC_TOKEN_URI=https://<GITLAB HOST URL>/oauth/token
OIDC_USERINFO_URI=https://<GITLAB HOST URL>/oauth/userinfo
OIDC_USERNAME_CLAIM=username
OIDC_DISPLAY_NAME=GitLab
OIDC_SCOPES=openid email
```

Anyone who what's wrong?

**Comment by almereyda** [comment]
Hi!

I *believe* to remember, that the requested scopes `OIDC_SCOPES=openid email` need to be identical to the ones granted in OIDC, but I am not sure. You write:

> > I know the most permission at not necessary, but I just want to make sure its not the permission issues

Eventually there is a regression here?

Also the other two discussions that mention the same error #2942, #2606 revolve around network routing or eventually a self-signed certificate, that might be into play?

In [#3299#discussioncomment-2481398](https://github.com/outline/outline/discussions/3299#discussioncomment-2481398) we also found, that `OIDC_USERNAME_CLAIM` needs to read `nickname` for GitLab. https://github.com/outline/outline/discussions/3217#discussioncomment-2363966 on the contrary reports a working integration with `username`, but I'd give it a shot for a moment, just to try.

**Reply by bjun626** [reply]
Hi @almereyda,

> revolve around network routing  

^^^ This helped me

The initial error of `"Failed to obtain access token"` has been solve due to outline container is unable to access the <GITLAB HOST URL> (This is totally my fault). Its able to do that now. 

## Next Issues

After this I got another error which is similar to [#3299#discussioncomment-2461818](https://github.com/outline/outline/discussions/3299#discussioncomment-2461818)

`
{"error":"An email field was not returned in the profile parameter, but is required.","stack":"UnauthorizedError: An email field was not returned in the profile parameter, but is required.\n at AuthenticationError (/opt/outline/build/server/errors.js:33:34)\n at OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:78:47)\n at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:198:24\n at OAuth2Strategy._passportOauth.Strategy.userProfile (/opt/outline/build/server/routes/auth/providers/oidc.js:50:12)\n at processTicksAndRejections (node:internal/process/task_queues:96:5)","level":"error","message":"Error during authentication"}
`

For the OIDC_SCOPED I've set it to `OIDC_SCOPES=openid` and changed the GitLab scope to be only `openid` as this suggest [#3299#discussioncomment-2462018](https://github.com/outline/outline/discussions/3299#discussioncomment-2462018). I've also changed the `OIDC_USERNAME_CLAIM` to `nickname`.

I'm using Let's Encrypt cert, so there's no self-signed cert. Gitlab container and Outline container are able to ping each other local IP and URL.

## New Config

```
OIDC_CLIENT_ID=<REDACTED>
OIDC_CLIENT_SECRET=<REDACTED>
OIDC_AUTH_URI=https://<GITLAB HOST URL>/oauth/authorize
OIDC_TOKEN_URI=https://<GITLAB HOST URL>/oauth/token
OIDC_USERINFO_URI=https://<GITLAB HOST URL>/oauth/userinfo
OIDC_USERNAME_CLAIM=nickname
OIDC_DISPLAY_NAME=GitLab
OIDC_SCOPES=openid
```

![image](https://user-images.githubusercontent.com/34598848/164988156-5fa72bcb-3d76-42bf-a1a8-fdcbcbcc87ae.png)

**Comment by bjun626** [comment, marked as answer]
Found the solution

Basically you need OIDC_SCOPES for **email**

## ENV File
```diff
OIDC_CLIENT_ID=<REDACTED>
OIDC_CLIENT_SECRET=<REDACTED>
OIDC_AUTH_URI=https://<GITLAB HOST URL>/oauth/authorize
OIDC_TOKEN_URI=https://<GITLAB HOST URL>/oauth/token
OIDC_USERINFO_URI=https://<GITLAB HOST URL>/oauth/userinfo
OIDC_USERNAME_CLAIM=nickname
OIDC_DISPLAY_NAME=GitLab
- OIDC_SCOPES=openid
+ OIDC_SCOPES="openid email"
```

## GitLab Settings
![image](https://user-images.githubusercontent.com/34598848/164989409-4a444ddd-5eea-4a62-b330-d27c7605a41c.png)

**Reply by almereyda** [reply]
👏🏽 

Congratiulations.

This is what I thought.

I'm wondering myself why it's even working in our environment. 🤷🏽 

---

*Edit:* Now it appears you wouldn't need to put quotation marks around the `openid email` value for `OIDC_SCOPES`.

> From https://docs.docker.com/compose/env-file/#syntax-rules:
>
> > There is no special handling of quotation marks. This means that they are part of the VAL.

Happy that it's working for you anyway, too. Might be that the code in Outline, which makes use of that variable, is resilient enough to work around that.
